cpp/src/endpoint.cpp

```diff
@@ -148,14 +148,42 @@ void Endpoint::close()
     // the endpoint status is not UCS_OK
     closeMode = UCP_EP_CLOSE_MODE_FORCE;
   }
-  ucs_status_ptr_t status = ucp_ep_close_nb(_handle, closeMode);
-  if (UCS_PTR_IS_PTR(status)) {
-    auto worker = ::ucxx::getWorker(_parent);
-    while (ucp_request_check_status(status) == UCS_INPROGRESS)
-      if (!worker->isProgressThreadRunning()) worker->progress();
-    ucp_request_free(status);
-  } else if (UCS_PTR_STATUS(status) != UCS_OK) {
-    ucxx_error("Error while closing endpoint: %s", ucs_status_string(UCS_PTR_STATUS(status)));
+
+  auto worker = ::ucxx::getWorker(_parent);
+  ucs_status_ptr_t status;
+
+  if (worker->isProgressThreadRunning()) {
+    utils::CallbackNotifier callbackNotifierPre{false};
+    worker->registerGenericPre([this, &callbackNotifierPre, &status, closeMode]() {
+      status = ucp_ep_close_nb(_handle, closeMode);
+      callbackNotifierPre.store(true);
+    });
+    callbackNotifierPre.wait([](auto flag) { return flag; });
+
+    while (UCS_PTR_STATUS(status) == UCS_INPROGRESS) {
```

Comment:
Who can be updating `status`?

The pre-callback closes the endpoint, which produces some status. That "happens-before" this read because of the `wait` on the callbacknotifier.

Now we loop and fire off another callback which writes to `status` by calling `ucp_request_free` (which is the thing that will eventually turn `UCS_PTR_STATUS(status)` into `UCS_OK` I guess).

Since there is a `wait` in this loop, the `ucp_request_free` must "happen-before" the next iteration of the loop, so that read is safe. I wonder if there is some circumstance under which the write on line 168 can conflict with the read on line 163. I think yes, so we need:

```suggestion
    while (UCS_PTR_STATUS(status) == UCS_INPROGRESS) {
      utils::CallbackNotifier barrier{false};
      worker->registerGenericPre([&barrier]() { barrier.store(true); });
      barrier.wait([](auto flag) { return flag; });
```
To ensure correct sequencing.

Reply:
Otherwise the read of `status` in this thread can in theory migrate past the write on the progress thread (unlikely, but possible).

Reply:
> Now we loop and fire off another callback which writes to `status` by calling `ucp_request_free` (which is the thing that will eventually turn `UCS_PTR_STATUS(status)` into `UCS_OK` I guess).

No, `ucp_ep_close_nb` behaves the same way as any UCX request (like a tag request, for example). What will happen is that it will either return an actual status if it completed immediately, otherwise it will return a pointer to a request. If it completed immediately we're done, and when we check if it's still in progress it will just never enter the loop. `UCS_PTR_STATUS(status)` will check what is the status of the underlying request, if it completed immediately it will be either `UCS_OK` or `UCS_ERR_*`, if not the request is still `UCS_INPROGRESS` which will cause us to enter the loop. At this point, we have to wait for the request to complete, so while `UCS_PTR_IS_PTR(status)` (meaning it's still a request pointer) we need then to query again whether the request completed or not in line 167 `auto s = ucp_request_check_status(status)`, and if it's not `UCS_INPROGRESS` anymore (it completed) we can then free the request pointer, which won't modify it. I did miss updating `status` with the new value returned by `s`.

`ucp_request_free` just releases a request pointer, if it is a request pointer, and it won't modify `status`. For a while I was myself confused as to the differences of `UCS_PTR_STATUS` and `ucp_request_check_status`, so I went asking for confirmation. `UCS_PTR_STATUS` will basically return the status if the argument is a `ucs_status_t` value, and if it's a pointer it will return `UCS_INPROGRESS`, whereas `ucs_request_check_status` will check whether the request has actually completed and then return its final status, or `UCS_INPROGRESS` if it didn't.Therefore, the only place where we should assign it is after we've checked and released `status` (to update that it is _not_ a pointer anymore but a status now). This should now be fixed via https://github.com/rapidsai/ucxx/pull/78/commits/0260a50da0d010e0a011a682a5dae3dc7f840755 .